bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+                        return error!("Multiple keys are not allowed without pattern")
+                    }
+                },
+            };
+            for kn in keys_names {
+                info!(
+                    "GET key:\"{:?}\" from  \"{}\" to file \"{}\"",
+                    kn.key, addr, &kn.name
+                );
+                get(kn.key, &kn.name, addr.clone()).await; // i dont like this clone()...
             }
-            "get" => {
-                let addr = sub_mathes
-                    .value_of("uri")
-                    .expect("has default value")
-                    .parse()
-                    .expect("wrong format of url");
-                info!("GET key:\"{:?}\" from  \"{}\"", key, addr);
-                get(key, addr).await;
+        }
+        EXISTS_SC => {
+            match app_args.key_pattern.unwrap() {
+                KeyPattern::Single(val) => {
+                    let key = get_key_value(val, app_args.keysize);
+                    exist(vec![key], addr).await;
+                }
+                KeyPattern::Range(range) => {
+                    let keysize = app_args.keysize;
+                    let keys = range.map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
+                KeyPattern::Multiple(vec) => {
+                    let keysize = app_args.keysize;
+                    let keys = vec.into_iter().map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
             }
-            _ => {}
         }
+        _ => unreachable!(),
     }
 }
 
-async fn put(key: Vec<u8>, size: usize, addr: Uri) {
-    let mut client = BobApiClient::connect(addr).await.unwrap();
+async fn prepare_put_from_pattern(re_path: &Regex, dir: &PathBuf, key_size: usize) -> Vec<KeyName> {
+    let mut dir_iter = read_dir(dir).await.unwrap();
 
-    let timestamp = SystemTime::now()
-        .duration_since(UNIX_EPOCH)
-        .expect("msg: &str")
-        .as_secs();
-    let meta = BlobMeta { timestamp };
-    let blob = Blob {
-        data: vec![1; size],
-        meta: Some(meta),
-    };
-    let message = PutRequest {
-        key: Some(BlobKey { key }),
-        data: Some(blob),
-        options: None,
-    };
-    let put_req = Request::new(message);
+    let mut keys_names = Vec::new();
+    while let Some(entry) = dir_iter.next_entry().await.unwrap() {
+        let file = entry.path();
+            let name = file.to_str().unwrap().to_owned();
+            if let (false, Some(cap)) = (file.is_dir(), re_path.captures(&name)) {
+                let key = get_key_value(cap[1].parse().unwrap(), key_size);
+                keys_names.push(KeyName { key, name });
+            }
+    }
+    keys_names
+}
 
-    let res = client.put(put_req).await;
-    info!("{:#?}", res);
+fn prepare_get(filenames: Vec<(u64, String)>, key_size: usize) -> Vec<KeyName> {
```

Comment:
It is better to convert values lazily without vector allocation

Reply:
Done: [bcd4957](https://github.com/qoollo/bob/pull/562/commits/bcd49574ee4b049185631c2cd9663a183e813b8b)